### PR TITLE
Remove type instabilities in the time stepping code

### PR DESCRIPTION
**Pull request by simone-silvestri**
There are several type instabilities in our launch! function, which lead to outrageous memory allocation and hurt performance on distributed architectures where GC fires and synchronizes all the workers.
This is caused by the fact that `Kernel` type we construct is not inferrable, mainly because the worksize is not inferrable

This PR is a first pass to remove type instabilities:
-  the active_cell_maps route was the most egregious. By switching to a DynamicSize we do not have to compile a kernel with a StaticSize of a runtime value (length(active_cell_maps)) we have a concretely typed kernel. This does not hurt us in performance since the dynamic size is once again computed with `length(map)`.
- the :xyz path and the KernelParameters path were also using StaticSize of runtime values (the grid size), so the fix is to make the grid size a compile-side constant by embedding it as a parameter in the grid

cc @jackdfranklin I think this might solve the distributed problem (and generally remove allocations and speed up the code)

EDIT:

I have also corrected dangerous patters that would loop over non-type stable tuples and a genuine bug in the `getproperty` of `model.clock` where clock was saved in the model as `Clock{T}` while `Clock` has 5 type parameters so it could never be inferred cascading dynamic runtime evaluation everywhere.
The results at the moment are:

### On main:
```julia
julia> grid = LatitudeLongitudeGrid(size = (250, 250, 100), latitude=(30.2, 46), halo = (5, 5, 5), longitude=(-15, 36), z = (-2000, 0));

julia> grid = ImmersedBoundaryGrid(grid, GridFittedBottom((x, y) -> -1000); active_cells_map = true);

julia> model = HydrostaticFreeSurfaceModel(grid; free_surface=SplitExplicitFreeSurface(grid; substeps = 100), momentum_advection=WENO(), tracer_advection=WENO(), tracers = (:T, :S));

julia> for _ in 1:10
       time_step!(model, 1)
       end

julia> @ballocated time_step!(model, 1)
5322384
```

### On this PR:
```julia
julia> grid = LatitudeLongitudeGrid(size = (250, 250, 100), latitude=(30.2, 46), halo = (5, 5, 5), longitude=(-15, 36), z = (-2000, 0));

julia> grid = ImmersedBoundaryGrid(grid, GridFittedBottom((x, y) -> -1000); active_cells_map = true);

julia> model = HydrostaticFreeSurfaceModel(grid; free_surface=SplitExplicitFreeSurface(grid; substeps = 100), momentum_advection=WENO(), tracer_advection=WENO(), tracers = (:T, :S));

julia> for _ in 1:10
       time_step!(model, 1)
       end

julia> @ballocated time_step!(model, 1)
11680

```

Those extra allocations are all now kernel launches. We allocate exactly 32B per kernel launch independently on the kernel or the number of arguments (it was kBs in main and scaling with the number of arguments), so the extra allocations now are all in:
```
4608   step_split_explicit_free_surface.jl:167 iterate_split_explicit_in_halo!
3976   multi_region_transformation.jl:86 #apply_regionally!#46
672    fill_halo_regions.jl:37 #fill_halo_event!#27
472    field_tuples.jl:32 flattened_unique_values
256    fill_halo_regions_normal_flow.jl:11 #fill_halo_event!#33
232    field_tuples.jl:27 flattened_unique_values
192    cache_hydrostatic_free_surface_tendencies.jl:71 cache_prognostic_field_tendency!
192    mask_immersed_field.jl:58 mask_immersed_field!
128    field_tuples.jl:38 tuplify
64     barotropic_split_explicit_corrector.jl:26 compute_barotropic_mode!
64     hydrostatic_free_surface_ab2_step.jl:235 ab2_step_tracer!
64     compute_hydrostatic_free_surface_tendencies.jl:130 launch_tracer_tendencies!
```

### CODE REVIEW

**Comment by simone-silvestri** [comment]
I will also add some tests for both hydrostatic and nonhydrostatic model to check that the allocations during time stepping are below a certain threshold, so that we make sure that we do not spike the allocations in future development

**Comment by giordano** [comment]
Today I'm a bit busy, will have a look later, but it'd be good to have @jackdfranklin test this when he has a chance!

File: src/DistributedComputations/halo_communication.jl [hunk 1]
```diff
@@ -74,29 +74,30 @@ end
 ##### Filling halos for halo communication boundary conditions
 #####
 
+# The communication buffers and the boundary-condition `args` (e.g. `clock`, `fields(model)`) are bundled into a
+# single tuple that is threaded by reference through the whole distributed halo machinery.
```
Comment (simone-silvestri) [inline src/DistributedComputations/halo_communication.jl]:
```suggestion
```

File: src/DistributedComputations/halo_communication.jl [hunk 2]
```diff
@@ -167,19 +176,19 @@ end
 cooperative_wait(req::MPI.Request)            = MPI.Waitall(req)
 cooperative_waitall!(req::Array{MPI.Request}) = MPI.Waitall(req)
 
-# Fallback: for serial boundary conditions fall back to `fill_halo_event!` but prune out the additional `buffers`
-# argument used only for distributed halo-filling boundary conditions
-distributed_fill_halo_event!(c, kernel!, bcs, loc, grid::DistributedGrid, buffers, args...; kwargs...) =
-    fill_halo_event!(c, kernel!, bcs, loc, grid, args...; kwargs...)
+# Fallback: for serial boundary conditions fall back to `fill_halo_event!`, splatting the boundary-condition args
+# (`args` is `(buffers, bc_args...)`; `Base.tail` prunes the distributed-only `buffers`).
```
Comment (simone-silvestri) [inline src/DistributedComputations/halo_communication.jl]:
```suggestion
# Fallback: for serial boundary conditions `Base.tail` prunes the distributed-only `buffers`.
```

**Comment by simone-silvestri** [comment]
I think the biggest gain is this:

| Benchmark | PR (pts/s) | Main (pts/s) | Change |
|-----------|------------|--------------|--------|
EarthOcean_tripolar_180x90x50_F64_WENOVectorInvariantDefault_WENO7_CATKE_2tr | 35185749.922 | 23818648.424 | +47.7% |

Not sure if it is correct since with small grids the performance metric is a bit noisier, but I would expect a reduction in allocation to improve only small cases on the GPU

**Comment by giordano** [comment]
What's up with `EarthOcean_tripolar_180x90x50_F64_WENOVectorInvariantDefault_WENO7_CATKE_2tr`?

**Comment by simone-silvestri** [comment]
> What's up with `EarthOcean_tripolar_180x90x50_F64_WENOVectorInvariantDefault_WENO7_CATKE_2tr`?

The explanation I hope is: on main the allocations dominate the runtime given this benchmark is just too small to saturate the GPU, so reducing the allocations speeds up the case. The same speedup is not seen in other cases that mask GC through more GPU compute. However we can show it definitively only with nsys profiles. I could produce a couple and show them here.

What I am more surprised of and cannot explain is the regression in performance of the `w` compute kernel

| Kernel | Median (ms) | Main (ms) | Change | Instances | Avg (ms) | Min (ms) | Max (ms) |
|--------|------------|-----------|--------|-----------|----------|----------|----------|
| `gpu__compute_w_from_continuity_` | 0.346 | 0.317 | +9.2% | 633 | 0.346 | 0.331 | 0.362 |

**Comment by simone-silvestri** [comment]
This PR seems to be the right way forward but as I continued, it adds a lot of different fixes in different places in the code. I will split it up in (probably) 5 PRs that will tackle one instability / allocation at the time.

One first one (which was still absent here) is #5737 

Btw, the new allocation count is now
```julia
julia> @time time_step!(model, 1)
  2.896576 seconds (2 allocations: 64 bytes)
```

**Comment by jackdfranklin** [comment]
This looks promising, especially since the smaller grid has such a large speedup. That should transfer to the distributed case. I will start to profile this branch (on single GPU and distributed) so that we can see how the GC has changed.

File: src/Utils/kernel_launching.jl [hunk 3]
```diff
@@ -602,3 +585,50 @@ end
...
+##### through `launch!`, so only a global override catches them all. The body is KA's `__run` verbatim, plus
+##### the single `let` marked below, which keeps the args box out of the single-threaded path. NOTE: this is a
+##### method overwrite of a KA function and therefore breaks Oceananigans precompilation; it is a deliberate,
+##### temporary measure until the `let` fix lands upstream in KernelAbstractions.
+#####
+
+import KernelAbstractions: __run
+using KernelAbstractions: __thread_run
+
+function __run(obj, ndrange, iterspace, args, dynamic, static_threads)
```
Comment (jackdfranklin) [inline src/Utils/kernel_launching.jl:606]: This overwrite causes the following warning:
```
WARNING: Method definition __run(Any, Any, Any, Any, Any, Any) in module KernelAbstractions at /glade/work/jfranklin/.julia/packages/KernelAbstractions/ecO4B/src/cpu.jl:98 overwritten in module Utils at /glade/u/home/jfranklin/CliMA/Oceananigans.jl/src/Utils/kernel_launching.jl:606.
│  ERROR: Method overwriting is not permitted during Module precompilation. Use `__precompile__(false)` to opt-out of precompilation.
```
I think we should probably avoid doing this, and instead change the way we launch the kernels (i.e. cache them). The calls to kernels that don't go through launch can (and maybe should) be changed.

**Comment by jackdfranklin** [comment]
@simone-silvestri Could you document all the type instabilities that you have found? Maybe create an issue for it, and then create sub-issues for each instability.

**Comment by giordano** [comment]
One was #5738

File: src/Utils/kernel_launching.jl (hunk 3 above)
Comment (simone-silvestri) [reply src/Utils/kernel_launching.jl:606]: I think this is a PR to do in KernelAbstractions. Caching the kernels will not help because the allocation is in the KA `__run` function that boxes the args and allocates heap memory

Comment (simone-silvestri) [reply src/Utils/kernel_launching.jl:606]: btw this branch is not to be merged, it is only a PR that collects all the optimizations and fixes which I am planning to add one by one in different PR, for example: #5737, #5739 and #5740

File: src/Grids/orthogonal_spherical_shell_grid.jl [hunk 4]
```diff
@@ -1,6 +1,6 @@
 const AHCG = AbstractHorizontallyCurvilinearGrid
 
-struct OrthogonalSphericalShellGrid{FT, TX, TY, TZ, Z, Map, CC, FC, CF, FF, Arch, FT2} <: AHCG{FT, TX, TY, TZ, Z, Arch}
+struct OrthogonalSphericalShellGrid{FT, TX, TY, TZ, Z, Map, CC, FC, CF, FF, Arch, FT2, Sz} <: AHCG{FT, TX, TY, TZ, Z, Arch}
```
Comment (glwagner) [inline src/Grids/orthogonal_spherical_shell_grid.jl:3]: What is the rationale for making this the final type parameter, vs another position?

File: src/Grids/orthogonal_spherical_shell_grid.jl [hunk 5]
```diff
@@ -47,7 +47,7 @@ function OrthogonalSphericalShellGrid{FT, TX, TY, TZ}(architecture::Arch,
                                                   radius :: FT2,
                                                   conformal_mapping :: Map) where {TX, TY, TZ, FT, Z, Map,
                                                                                    CC, FC, CF, FF, Arch, FT2}
-    return OrthogonalSphericalShellGrid{FT, TX, TY, TZ, Z, Map, CC, FC, CF, FF, Arch, FT2}(architecture,
+    return OrthogonalSphericalShellGrid{FT, TX, TY, TZ, Z, Map, CC, FC, CF, FF, Arch, FT2, GridSize{Int(Nx), Int(Ny), Int(Nz), Int(Hx), Int(Hy), Int(Hz)}}(architecture,
```
Comment (glwagner) [inline src/Grids/orthogonal_spherical_shell_grid.jl:50]: can you convert to `Int` within the `GridSize` constructor?

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/initialize_split_explicit_substepping.jl [hunk 6]
```diff
@@ -33,9 +33,11 @@ function initialize_free_surface_state!(free_surface, baroclinic_timestepper, ti
...
     initialize_free_surface_timestepper!(timestepper, η, U, V)
 
-    for field in free_surface.filtered_state
-        fill!(field, 0)
-    end
+    fill!(free_surface.filtered_state.η̅, 0)
+    fill!(free_surface.filtered_state.U̅, 0)
+    fill!(free_surface.filtered_state.V̅, 0)
+    fill!(free_surface.filtered_state.Ũ, 0)
+    fill!(free_surface.filtered_state.Ṽ, 0)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/initialize_split_explicit_substepping.jl:40]: usually better to launch a single kernel for this

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/initialize_split_explicit_substepping.jl [hunk 7]
```diff
@@ -57,9 +59,11 @@ function initialize_free_surface_state!(free_surface, baroclinic_ts::SplitRungeK
...
     initialize_free_surface_timestepper!(barotropic_ts, η, U, V)
 
-    for field in free_surface.filtered_state
-        fill!(field, 0)
-    end
+    fill!(free_surface.filtered_state.η̅, 0)
+    fill!(free_surface.filtered_state.U̅, 0)
+    fill!(free_surface.filtered_state.V̅, 0)
+    fill!(free_surface.filtered_state.Ũ, 0)
+    fill!(free_surface.filtered_state.Ṽ, 0)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/initialize_split_explicit_substepping.jl:66]: ditto as above

File: src/Models/HydrostaticFreeSurfaceModels/cache_hydrostatic_free_surface_tendencies.jl [hunk 8]
```diff
@@ -100,21 +98,27 @@ properly handle mutable vertical coordinates (z-star). Velocities and free surfa
...
-        Ψⁿ = model_fields[name]
-        if name ∈ keys(model.tracers) # Tracers are stored with the grid scaling
-            launch!(arch, grid, :xyz, _cache_tracer_fields!, Ψ⁻, grid, Ψⁿ)
-        else # Velocities and free surface are stored without the grid scaling
-            parent(Ψ⁻) .= parent(Ψⁿ)
-        end
+    if name ∈ keys(model.tracers) # Tracers are stored with the grid scaling
+        launch!(architecture(grid), grid, :xyz, _cache_tracer_fields!, Ψ⁻, grid, Ψⁿ)
+    else # Velocities and free surface are stored without the grid scaling
+        parent(Ψ⁻) .= parent(Ψⁿ)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/cache_hydrostatic_free_surface_tendencies.jl:121]: is there a reason not to do this with one kernel?

File: src/Models/interleave_communication_and_computation.jl [hunk 9]
```diff
@@ -24,18 +24,23 @@ complete_communication_and_compute_buffer!(model, grid, arch) = nothing
...
-interior_tendency_kernel_parameters(arch, grid) = :xyz # fallback
+@inline interior_tendency_kernel_parameters(arch, grid) = KernelParameters(worksize(grid), map(zero, worksize(grid))) # fallback
 
-function interior_tendency_kernel_parameters(arch::AsynchronousDistributed, grid)
-    Rx, Ry, _ = arch.ranks
+@inline local_dimension(T) = T == Bounded           || 
+                             T == Periodic          || 
+                             T == Flat              || 
+                             T == RightFaceFolded   || 
+                             T == RightCenterFolded
```
Comment (glwagner) [inline src/Models/interleave_communication_and_computation.jl:33]: should these be `===`?

File: test/test_memory_allocation.jl [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+function time_step_allocations(model, Δt; samples=10)
+    # Compile first
+    time_step!(model, Δt)
+    time_step!(model, Δt)
+    time_step!(model, Δt)
+    time_step!(model, Δt)
+    time_step!(model, Δt)
+
+    # Then measure allocation
+    return minimum(@allocated(time_step!(model, convert(eltype(model.grid), Δt))) for _ in 1:samples)
```
Comment (glwagner) [inline test/test_memory_allocation.jl:44]: write this in a sensible way

File: test/test_memory_allocation.jl [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+using Oceananigans.TurbulenceClosures: CATKEVerticalDiffusivity
+using Oceananigans.DistributedComputations: @handshake
+using Oceananigans.Utils: pretty_filesize
+
+function allocation_grid(arch, FT=Float64; immersed_mode, size, extent=(1, 1, 1), halo=(7, 7, 7), topology=(Periodic, Periodic, Bounded))
+
+    grid = RectilinearGrid(arch, FT; size, extent, halo, topology)
+    immersed_mode == :flat && return grid
+
+    bump(x, y) = - extent[3] * (1 - convert(FT, 0.3) * exp(-((x - extent[1]/2)^2 + (y - extent[2]/2)^2)))
```
Comment (glwagner) [inline test/test_memory_allocation.jl:13]:
```suggestion
    Lx, Ly, Lz = extent
    bump(x, y) = - Lz * (1 -0.3 * exp(-((x - Lx/2)^2 + (y - Ly/2)^2)))
```

File: src/Grids/orthogonal_spherical_shell_grid.jl (hunk 4 above)
Comment (simone-silvestri) [reply src/Grids/orthogonal_spherical_shell_grid.jl:3]: This is used only in `Base.size` so it is good to keep it at the end, given that it will be never used for dispatch

File: src/Models/HydrostaticFreeSurfaceModels/cache_hydrostatic_free_surface_tendencies.jl (hunk 8 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/cache_hydrostatic_free_surface_tendencies.jl:121]: I guess this is less code and avoids any possible kernel issue

**Comment by simone-silvestri** [comment]
Everything here has been implemented in separate, smaller PRs
